### DISCUSSION TITLE
terminate cmd completion status and IKE tunnel down event sequence.

### DISCUSSION THREAD

**Discussion by digambaringale333** [strongswan/strongswan#412] [Q&A]
Hi,
We are using davici code in our application to create secure tunnel between two machines, while working on terminate path we did some POC's with terminate API and have some queries based on our observations. We have already registered for up/down events with VICI before issuing terminate request.

1. We issued terminate with force flag and default timeout (which is infinite) and observed that we get the IKE down event first and then command completion status => Is this the guaranteed behavior that we will get tunnel down event first and then command completion status always?
2. When we terminate tunnel with force flag with default timeout, we could see that tunnel gets terminated in all cases (error cases too) => is this the guaranteed behavior that tunnel will get terminated successfully in all cases?

We want confirm this behavior, if we can rely on above observations to make some design decision's.

Thanks for your help, I appreciate it!

**Comment by tobiasbrunner** [comment]
>  Is this the guaranteed behavior that we will get tunnel down event first and then command completion status always?

Only if you use a single VICI connection for both. While the down event is triggered before the blocking terminate call returns (i.e. the event message should be queued on the VICI socket before the command reply), the listeners for the former are notified asynchronously. So with multiple connections, they could get the event message after the terminate call returns if there is a delay somewhere (e.g. due to other events or just because of how threads are scheduled).

>    is this the guaranteed behavior that tunnel will get terminated successfully in all cases?

Yes, the IKE_SA is just forcefully destroyed locally (a single delete is sent to the peer if the state is ESTABLISHED, but the daemon won't wait for a response or retransmit it).

**Reply by digambaringale333** [reply]
Thank you very much Tobais for the prompt reply, we appreciate it!!!

**Reply by digambaringale333** [reply]
Hi Tobias,
I have one follow up observations/query on the above mentioned behavior. We are using only a single connection to communicate with VICI.

Terminate API config:
1. I have configured terminate timeout as "0" (infinite)
2. Added force flag.

Experiment:
1. I created a IPSec tunnel between two machines(Host-1, Host-2).
2. I issued terminate cmd on Host-1, on host-1 I got the tunnel down event first and then got the command completion - Looks good :)
3. When I issued terminate cmd on Host-1, immediately(after few ms) I issued terminate command on Host-2. In this case on Host-2 I got the command response first "no matching SAs to terminate found" (Strongswan might have terminated tunnel on Host-2, because peer terminated the tunnel from his side) and then I got the tunnel down event.

Will you please help us to understand this behavior?

**Reply by tobiasbrunner** [reply]
What don't you understand?

**Reply by digambaringale333** [reply]
Hi, As we discussed earlier we were assuming that we will always get down event first and then command completion status(because we are using single connection) but as I mentioned in above step-3 we are getting tunnel down event after command completion status. Why are we getting command completion first and then event?

**Reply by tobiasbrunner** [reply]
Because it's triggered by the delete from the peer and not by the terminate command.
